src/Symfony/Component/AssetMapper/Factory/CachedMappedAssetFactory.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AssetMapper\Factory;
+
+use Symfony\Component\AssetMapper\MappedAsset;
+use Symfony\Component\Config\ConfigCache;
+use Symfony\Component\Config\Resource\FileResource;
+
+class CachedMappedAssetFactory implements MappedAssetFactoryInterface
+{
+    public function __construct(
+        private MappedAssetFactoryInterface $innerFactory,
+        private string $cacheDir,
+    ) {
+    }
+
+    public function createMappedAsset(string $logicalPath, string $sourcePath): ?MappedAsset
+    {
+        $cachePath = $this->getCacheFilePath($logicalPath, $sourcePath);
+        $configCache = new ConfigCache($cachePath, true);
+
+        if ($configCache->isFresh()) {
+            return unserialize(file_get_contents($cachePath));
+        }
+
+        $mappedAsset = $this->innerFactory->createMappedAsset($logicalPath, $sourcePath);
+
+        if (!$mappedAsset) {
+            return null;
+        }
+
+        $resources = [new FileResource($mappedAsset->getSourcePath())];
+        foreach ($mappedAsset->getDependencies() as $dependency) {
+            if (!$dependency->contentDependency) {
+                continue;
+            }
+
+            $resources[] = new FileResource($dependency->asset->getSourcePath());
+        }
+        $configCache->write(serialize($mappedAsset), $resources);
```

Comment:
Given that the MappedAsset object contains AssetDependency objects that contain other MappedAsset objects, what happens when you serialize and unserialize those ?

Also, what is the impact of unserializing a MappedAsset with dependencies that will have MappedAsset objects not known by the AssetMapper internal cache (while potentially referring to the same assets) and also different from MappedAsset created for dependencies of other cached MappedAsset ?

Reply:
> Given that the MappedAsset object contains AssetDependency objects that contain other MappedAsset objects, what happens when you serialize and unserialize those ?

It seems to work fine, in practice. Both `MappedAsset` and `AssetDependency` are objects that hold scalars. For example, here is a dump of one situation (dumping the result of the `unserialize` in the cached factory) in a test project:

<img width="735" alt="Screenshot 2023-05-10 at 1 32 25 PM" src="https://github.com/symfony/symfony/assets/121003/2fb08f63-4726-4b07-ae76-2eea5706c9bd">

What concerns do you have?

> Also, what is the impact of unserializing a MappedAsset with dependencies that will have MappedAsset objects not known by the AssetMapper internal cache (while potentially referring to the same assets) and also different from MappedAsset created for dependencies of other cached MappedAsset ?

I think I understand what you're referring to. So, for example:

* `asset1.css` has a dependency on `asset2.css`. When something asks for the `asset1.css` asset, its cache is fresh, so we get back `MappedAsset` instances from the cache for both `asset1.css` and `asset2.css`.
* Meanwhile, `asset3.css` also has a dependency on `asset2.css`. When something asks for the `asset3.css` asset, its cache is NOT fresh, so we get back FRESH `MappedAsset `instances for both `asset3.css` and `asset2.css`.

So, at this moment, there are two distinct `asset2.css` `MappedAsset` objects floating around the system. Their data is equal, but the objects are not identical. Other than being wasteful, I don't think it's a problem. `MappedAsset` objects are "read-only" informational objects. Well, except for `MappedAsset::addDependency()`. But, in practice, once a `MappedAsset` is returned from `AssetMapper`, it is intended to be read-only: its data is entirely set *during* the creation process (we could even "enforce" this by adding some flag to "freeze" the `MappedAsset` and thus reject calling `addDependency()`, but that might be overkill).

However, if we're concerned about this, we *could*, "refresh" the dependencies after fetching a cached `MappedAsset` - e.g. loop over `$mappedAsset->getDependencies()`... grab the `->logicalName` + `->getSourcePath()` of each, call `$this->createMappedAsset($logicalName, $sourcePath)` on each, then "replace" the `MappedAsset` on each dependency.

tl;dr I don't see any real problem, but I could add extra code to be absolutely sure.